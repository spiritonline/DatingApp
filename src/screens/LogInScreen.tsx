import { useState } from 'react';
import { useColorScheme, View, Alert, ScrollView, KeyboardAvoidingView, Platform } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { signInWithEmailAndPassword } from '@firebase/auth';
import { auth } from '../services/firebase';
import { AuthNavigationProp } from '../navigation/types';
import styled from 'styled-components/native';
import { SafeAreaView } from 'react-native-safe-area-context';
import AuthHeader from '../components/auth-components/AuthHeader';

export default function LogInScreen() {
  const colorScheme = useColorScheme();
  const navigation = useNavigation<AuthNavigationProp>();
  const isDark = colorScheme === 'dark';

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  const [errors, setErrors] = useState({
    email: '',
    password: '',
  });

  const validateForm = () => {
    let isValid = true;
    const newErrors = {
      email: '',
      password: '',
    };

    // Email validation
    if (!email) {
      newErrors.email = 'Email is required';
      isValid = false;
    } else if (!/\S+@\S+\.\S+/.test(email)) {
      newErrors.email = 'Email is invalid';
      isValid = false;
    }

    // Password validation
    if (!password) {
      newErrors.password = 'Password is required';
      isValid = false;
    }

    setErrors(newErrors);
    return isValid;
  };

  const handleLogIn = async () => {
    if (!validateForm()) return;

    setIsLoading(true);
    try {
      await signInWithEmailAndPassword(auth, email, password);
      navigation.navigate('MainFeed');
    } catch (error: any) {
      console.error('Error logging in:', error);
      Alert.alert(
        'Login Failed',
        error.message || 'Failed to log in. Please check your credentials and try again.'
      );
    } finally {
      setIsLoading(false);
    }
  };

  const handleGoogleLogIn = () => {
    // Implementation would go here in a real app
    Alert.alert('Google Login', 'Google Login would be implemented here');
  };

  const handleAppleLogIn = () => {
    // Implementation would go here in a real app
    Alert.alert('Apple Login', 'Apple Login would be implemented here');
  };

  return (
    <Container isDark={isDark}>
      <KeyboardAvoidingView
        behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
        style={{ flex: 1, width: '100%' }}
      >
        <ScrollView 
          contentContainerStyle={{ flexGrow: 1 }}
          keyboardShouldPersistTaps="handled"
        >
          <AuthHeader title="Log In" onBack={() => navigation.goBack()} />
          <FormContainer>
            <InputContainer>
              <InputLabel isDark={isDark}>Email</InputLabel>
              <Input
                isDark={isDark}
                value={email}
                onChangeText={setEmail}
                placeholder="Enter your email"
                placeholderTextColor="#999"
                keyboardType="email-address"
                autoCapitalize="none"
                autoComplete="email"
                testID="email-input"
                accessibilityLabel="Email input"
                accessibilityHint="Enter your email address"
              />
              {errors.email ? <ErrorText>{errors.email}</ErrorText> : null}
            </InputContainer>

            <InputContainer>
              <InputLabel isDark={isDark}>Password</InputLabel>
              <Input
                isDark={isDark}
                value={password}
                onChangeText={setPassword}
                placeholder="Enter your password"
                placeholderTextColor="#999"
                secureTextEntry
                testID="password-input"
                accessibilityLabel="Password input"
                accessibilityHint="Enter your password"
              />
              {errors.password ? <ErrorText>{errors.password}</ErrorText> : null}
            </InputContainer>

            <ForgotPasswordText 
              onPress={() => Alert.alert('Forgot Password', 'Password reset would be implemented here')}
              accessibilityRole="button"
              accessibilityLabel="Forgot Password"
              accessibilityHint="Reset your password"
            >
              Forgot Password?
            </ForgotPasswordText>

            <LoginButton 
              onPress={handleLogIn} 
              disabled={isLoading}
              testID="login-button"
              accessibilityRole="button"
              accessibilityLabel="Log In Button"
              accessibilityHint="Log in with your email and password"
            >
              <ButtonText>{isLoading ? 'Logging In...' : 'Log In'}</ButtonText>
            </LoginButton>

            <Divider isDark={isDark}>
              <DividerLine isDark={isDark} />
              <DividerText isDark={isDark}>or log in with</DividerText>
              <DividerLine isDark={isDark} />
            </Divider>

            <SocialButtonsContainer>
              <SocialButton 
                onPress={handleGoogleLogIn}
                testID="google-login-button"
                accessibilityRole="button"
                accessibilityLabel="Log in with Google"
              >
                <SocialButtonText isDark={isDark}>Google</SocialButtonText>
              </SocialButton>

              <SocialButton 
                onPress={handleAppleLogIn}
                testID="apple-login-button"
                accessibilityRole="button"
                accessibilityLabel="Log in with Apple"
              >
                <SocialButtonText isDark={isDark}>Apple</SocialButtonText>
              </SocialButton>
            </SocialButtonsContainer>

            <BottomText>
              Don't have an account?{' '}
              <LinkText 
                onPress={() => navigation.navigate('SignUp')}
                accessibilityRole="link"
                accessibilityLabel="Sign Up"
                accessibilityHint="Navigate to sign up screen"
              >
                Sign Up
              </LinkText>
            </BottomText>
          </FormContainer>
        </ScrollView>
      </KeyboardAvoidingView>
    </Container>
  );
}

import { ThemeProps } from '../utils/styled-components';

const Container = styled(SafeAreaView)<ThemeProps>`
  flex: 1;
  background-color: ${(props: ThemeProps) => (props.isDark ? '#121212' : '#ffffff')};
`;

const FormContainer = styled.View`
  flex: 1;
  padding: 20px;
  align-items: center;
`;

const InputContainer = styled.View`
  width: 100%;
  margin-bottom: 16px;
`;

const InputLabel = styled.Text<ThemeProps>`
  font-size: 14px;
  margin-bottom: 8px;
  color: ${(props: ThemeProps) => (props.isDark ? '#e0e0e0' : '#333')};
`;

const Input = styled.TextInput<ThemeProps>`
  width: 100%;
  padding: 16px;
  border-radius: 8px;
  background-color: ${(props: ThemeProps) => (props.isDark ? '#2c2c2c' : '#f5f5f5')};
  color: ${(props: ThemeProps) => (props.isDark ? '#ffffff' : '#000000')};
  border: 1px solid ${(props: ThemeProps) => (props.isDark ? '#444' : '#ddd')};
`;

const ErrorText = styled.Text`
  color: #ff3b30;
  font-size: 12px;
  margin-top: 4px;
`;

const ForgotPasswordText = styled.Text`
  color: #FF6B6B;
  font-size: 14px;
  align-self: flex-end;
  margin-bottom: 20px;
`;

interface ButtonProps {
  disabled?: boolean;
}

const LoginButton = styled.TouchableOpacity<ButtonProps>`
  width: 100%;
  padding: 16px;
  border-radius: 8px;
  background-color: #FF6B6B;
  align-items: center;
  opacity: ${(props: ButtonProps) => (props.disabled ? 0.7 : 1)};
`;

const ButtonText = styled.Text`
  color: white;
  font-size: 16px;
  font-weight: bold;
`;

const Divider = styled.View<ThemeProps>`
  flex-direction: row;
  align-items: center;
  width: 100%;
  margin: 30px 0;
`;

const DividerLine = styled.View<ThemeProps>`
  flex: 1;
  height: 1px;
  background-color: ${(props: ThemeProps) => (props.isDark ? '#444' : '#ddd')};
`;

const DividerText = styled.Text<ThemeProps>`
  color: ${(props: ThemeProps) => (props.isDark ? '#888' : '#999')};
  padding: 0 10px;
  font-size: 12px;
`;

const SocialButtonsContainer = styled.View`
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
`;

const SocialButton = styled.TouchableOpacity`
  flex: 0.48;
  padding: 14px;
  border-radius: 8px;
  border: 1px solid #ddd;
  align-items: center;
`;

const SocialButtonText = styled.Text<ThemeProps>`
  font-size: 14px;
  color: ${(props: ThemeProps) => (props.isDark ? '#e0e0e0' : '#333')};
`;

const BottomText = styled.Text`
  margin-top: 30px;
  font-size: 14px;
  color: #999;
`;

const LinkText = styled.Text`
  color: #FF6B6B;
  font-weight: bold;
`;
