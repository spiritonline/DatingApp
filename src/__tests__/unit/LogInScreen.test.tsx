import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import LogInScreen from '../../screens/LogInScreen';
import { Alert } from 'react-native';

// Mock dependencies
const mockNavigate = jest.fn();
const mockGoBack = jest.fn();
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
      goBack: mockGoBack,
    }),
  };
});

// Mock Firebase auth
jest.mock('../../services/firebase', () => ({
  auth: {},
}));

jest.mock('firebase/auth', () => ({
  signInWithEmailAndPassword: jest.fn(() => Promise.resolve()),
}));

// Mock Alert
jest.spyOn(Alert, 'alert').mockImplementation(() => {});

describe('LogInScreen', () => {
  beforeEach(() => {
    mockNavigate.mockClear();
    mockGoBack.mockClear();
    jest.clearAllMocks();
  });

  it('renders correctly with all form fields', () => {
    const { getByText, getByTestId } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    
    // Check for essential elements
    expect(getByTestId('email-input')).toBeTruthy();
    expect(getByTestId('password-input')).toBeTruthy();
    expect(getByText('Forgot Password?')).toBeTruthy();
    expect(getByTestId('login-button')).toBeTruthy();
  });

  it('shows validation errors when form is submitted with empty fields', () => {
    const { getByText, getByTestId } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    
    const loginButton = getByTestId('login-button');
    fireEvent.press(loginButton);
    
    expect(getByText('Email is required')).toBeTruthy();
    expect(getByText('Password is required')).toBeTruthy();
  });

  it('navigates to MainFeed after successful login', async () => {
    const { getByTestId } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    
    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const loginButton = getByTestId('login-button');
    
    fireEvent.changeText(emailInput, 'test@example.com');
    fireEvent.changeText(passwordInput, 'password123');
    
    fireEvent.press(loginButton);
    
    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('MainFeed');
    });
  });

  it('navigates back when back button is pressed', () => {
    const { getByTestId } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    
    const backButton = getByTestId('back-button');
    fireEvent.press(backButton);
    
    expect(mockGoBack).toHaveBeenCalledTimes(1);
  });

  it('navigates to SignUp screen when Sign Up link is pressed', () => {
    const { getByText } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    
    const signUpLink = getByText('Sign Up');
    fireEvent.press(signUpLink);
    
    expect(mockNavigate).toHaveBeenCalledWith('SignUp');
  });

  it('shows alert when Forgot Password is pressed', () => {
    const { getByText } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    
    const forgotPasswordText = getByText('Forgot Password?');
    fireEvent.press(forgotPasswordText);
    
    expect(Alert.alert).toHaveBeenCalledWith('Forgot Password', 'Password reset would be implemented here');
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <NavigationContainer>
        <LogInScreen />
      </NavigationContainer>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
