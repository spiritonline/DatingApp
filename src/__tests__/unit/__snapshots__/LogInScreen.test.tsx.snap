// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogInScreen matches snapshot 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  isDark={false}
  style={
    {
      "backgroundColor": "#ffffff",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      color="#FF6B6B"
      size="large"
    />
  </View>
</RNCSafeAreaView>
`;
